lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,375 @@
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Mutex.h"
+#include "llvm/Support/RWMutex.h"
+
+#define DEBUG_TYPE "firrtl-lower-groups"
+
+using namespace circt;
+using namespace firrtl;
+
+class LowerGroupsPass : public LowerGroupsBase<LowerGroupsPass> {
+  /// Safely build a new module with a given namehint.  This handles geting a
+  /// lock to modify the top-level circuit.
+  FModuleOp buildNewModule(OpBuilder &builder, Location location,
+                           Twine namehint, SmallVectorImpl<PortInfo> &ports);
+
+  /// Function to process each module.
+  void runOnModule(FModuleOp moduleOp);
+
+  /// Entry point for the function.
+  void runOnOperation() override;
+
+  /// Indicates exclusive access to modify the circuitNamespace and the circuit.
+  llvm::sys::SmartMutex<true> *circuitMutex;
+
+  /// A circuit-level namespace.
+  CircuitNamespace *circuitNamespace;
+};
+
+/// Multi-process safe function to build a module in the circuit and return it.
+/// The name provided is only a namehint for the module---a unique name will be
+/// generated if there are conflicts with the namehint in the circuit-level
+/// namespace.
+FModuleOp LowerGroupsPass::buildNewModule(OpBuilder &builder, Location location,
+                                          Twine namehint,
+                                          SmallVectorImpl<PortInfo> &ports) {
+  llvm::sys::SmartScopedLock<true> instrumentationLock(*circuitMutex);
+  StringRef newModuleName = circuitNamespace->newName(namehint);
+  FModuleOp newModule = builder.create<FModuleOp>(
+      location, builder.getStringAttr(newModuleName),
+      ConventionAttr::get(builder.getContext(), Convention::Internal), ports,
+      ArrayAttr{});
+  SymbolTable::setSymbolVisibility(newModule, SymbolTable::Visibility::Private);
+  return newModule;
+};
+
+/// Process amodule to remove any groups it has.
+void LowerGroupsPass::runOnModule(FModuleOp moduleOp) {
+  LLVM_DEBUG({
+    llvm::dbgs() << "Module: " << moduleOp.getModuleName() << "\n";
+    llvm::dbgs() << "  Examining Groups:\n";
+  });
+
+  CircuitOp circuitOp = moduleOp->getParentOfType<CircuitOp>();
+  StringRef circuitName = circuitOp.getName();
+
+  // A set of ops that this pass creates.  This is used to check if certain
+  // operations should be handled specially, e.g., moved outside of their
+  // defining group.
+  DenseSet<Operation *> createdOps;
+
+  // Post-order traversal that expands a group into its parent. Foreach gorup
+  // found do the following:
+  //
+  // 1. Create and connect one ref-type output port for each value defined in
+  //    this group that drives an instance marked lowerToBind and move this
+  //    instance outside the group.
+  // 2. Create one input port for each value captured by this group.
+  // 3. Create a new module for this group and move the (mutated) body of this
+  //    group to the new module.
```

Comment:
Move! Nice :+1: .

Reply:
Thanks! This turned out to be a pretty slick way to do it.